### PR TITLE
Adding apis/v1alpha1/Destination; future sink and subscriber object.

### PR DESCRIPTION
**Pull request by n3wscott**
Adding apis/v1alpha1/Destination; future sink and subscriber object.

A followup will be to add a destination resolver object that will assemble the destination into a full usable url.

### CODE REVIEW

File: apis/destination.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+// Destination represents a target of an invocation over HTTP.
+type Destination struct {
+	// ObjectReference points to an Addressable.
+	*corev1.ObjectReference `json:",inline"`
+
+	// URI is for direct URI Designations.
+	URI *URL `json:"uri,omitempty"`
+
+	// Path is used with the resulting URL from Addressable ObjectReference or
+	// URI. Must start with `/`. Only OPatgh
```
Comment (lionelvillard) [inline apis/destination.go]: oPatgh?

File: apis/destination.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+type Destination struct {
+	// ObjectReference points to an Addressable.
+	*corev1.ObjectReference `json:",inline"`
+
+	// URI is for direct URI Designations.
+	URI *URL `json:"uri,omitempty"`
+
+	// Path is used with the resulting URL from Addressable ObjectReference or
+	// URI. Must start with `/`.
+	Path *string `json:"path,omitempty"`
```
Comment (mattmoor) [inline apis/destination.go]: Does this override the path in the resolved URL or augment it?

File: apis/destination.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import (
+	"context"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"net/url"
+	"strings"
+)
+
+// Destination represents a target of an invocation over HTTP.
+type Destination struct {
```
Comment (mattmoor) [inline apis/destination.go]: I'm wondering if this belongs in a versioned directory...

File: apis/v1alpha1/destination_test.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2017 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline apis/v1alpha1/destination_test.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: apis/destination.go (hunk 2 above)
Comment (n3wscott) [reply apis/destination.go]: I added a comment that the intent is to append them but it is up to the implementation of the resolver. Which will be added to pkg in a followup.

File: apis/destination.go (hunk 3 above)
Comment (n3wscott) [reply apis/destination.go]: not a bad idea. Moved to apis/v1alpha1 

I am not sure if we want to give it a name or just a apis with version. It felt more natural to have it `apis/v1alpha1`

File: apis/destination.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	if ref.Name == "" {
+		errs = errs.Also(ErrMissingField("name"))
+	}
+	if ref.APIVersion == "" {
+		errs = errs.Also(ErrMissingField("apiVersion"))
+	}
+	if ref.Kind == "" {
+		errs = errs.Also(ErrMissingField("kind"))
+	}
+
```
Comment (lionelvillard) [inline apis/destination.go]: what about checking the other fields are not set (FieldPath UID, etc..)?

Comment (n3wscott) [reply apis/destination.go]: I am not sure I want to limit them. But I do know I need these ones at least.

File: apis/destination.go (hunk 2 above)
Comment (mattmoor) [reply apis/destination.go]: Having the semantics be context specific feels somewhat antithetical to duck typing, does it not?

File: apis/v1alpha1/destination.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
+/*
```
Comment (mattmoor) [inline apis/v1alpha1/destination.go]: Generally there's a group name between `apis` and `v1alpha1`...

File: apis/destination.go (hunk 2 above)
Comment (n3wscott) [reply apis/destination.go]: This is not a duck, it is a field implementation.

File: apis/v1alpha1/destination.go (hunk 6 above)
Comment (n3wscott) [reply apis/v1alpha1/destination.go]: this is a generic component, but not a duck

**Pull request merged**